File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/pageobjectadvanced/Driver.java
package pageobjectadvanced;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.WebDriverWait;

public class Driver {
    private static WebDriverWait browserWait;
    private static WebDriver browser;

    public static WebDriver getBrowser() {
        if (browser == null) {
            throw new NullPointerException("The WebDriver browser instance was not initialized. You should first call the start() method.");
        }
        return browser;
    }

    public static void setBrowser(WebDriver browser) {
        Driver.browser = browser;
    }

    public static WebDriverWait getBrowserWait() {
        if (browserWait == null || browser == null) {
            throw new NullPointerException("The WebDriver browser wait instance was not initialized. You should first call the start() method.");
        }
        return browserWait;
    }

    public static void setBrowserWait(WebDriverWait browserWait) {
        Driver.browserWait = browserWait;
    }

    public static void startBrowser(BrowserType browserType, int defaultTimeout) {

        switch (browserType) {
            case FIREFOX:
                WebDriverManager.firefoxdriver().setup();
                setBrowser(new FirefoxDriver());
                break;
            case CHROME:
                //setBrowser(new ChromeDriver());
                break;
            case EDGE:
                //setBrowser(new EdgeDriver());
                break;
            default:
                break;
        }

        setBrowserWait(new WebDriverWait(getBrowser(), defaultTimeout));
    }

    public static void startBrowser(BrowserType browserType) {
        startBrowser(browserType, 30);
    }

    public static void startBrowser() {
        startBrowser(BrowserType.FIREFOX);
    }

    public static void stopBrowser() {
        getBrowser().quit();
        setBrowser(null);
        setBrowserWait(null);
    }
}

File: WebDriver-Series/TestProjectPageObjectsSingleton/Pages/SingletonBingMainPage/Driver.cs
﻿// <copyright file="Driver.cs" company="Automate The Planet Ltd.">
// Copyright 2021 Automate The Planet Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
// <author>Anton Angelov</author>
// <site>https://automatetheplanet.com/</site>

using System;
using OpenQA.Selenium;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Support.UI;

namespace TestProjectPageObjectsSingleton.Pages.SingletonBingMainPage
{
    public static class Driver
    {
        private static WebDriverWait _browserWait;

        private static IWebDriver _browser;

        public static IWebDriver GetBrowser()
        {
            if (_browser == null)
            {
                throw new NullReferenceException("The WebDriver browser instance was not initialized. You should first call the method Start.");
            }

            return _browser;
        }

        private static void SetBrowser(IWebDriver value)
        {
            _browser = value;
        }

        public static WebDriverWait GetBrowserWait()
        {
            if (_browserWait == null || _browser == null)
            {
                throw new NullReferenceException("The WebDriver browser wait instance was not initialized. You should first call the method Start.");
            }

            return _browserWait;
        }

        private static void SetBrowserWait(WebDriverWait value)
        {
            _browserWait = value;
        }

        public static void StartBrowser(BrowserType browserType = BrowserType.Firefox, int defaultTimeOut = 30)
        {
            switch (browserType)
            {
                case BrowserType.Firefox:
                    SetBrowser(new FirefoxDriver());
                    break;
                case BrowserType.InternetExplorer:
                    break;
                case BrowserType.Chrome:
                    break;
                default:
                    throw new ArgumentException("You need to set a valid browser type.");
            }

            SetBrowserWait(new WebDriverWait(GetBrowser(), TimeSpan.FromSeconds(defaultTimeOut)));
        }

        public static void StopBrowser()
        {
            GetBrowser().Quit();
            SetBrowser(null);
            SetBrowserWait(null);
        }
    }
}
